File: src/lib/cpp/DSSession.cpp
#include "DSSession.h"
#include "DSCgi.h"
#include <stdlib.h>

#include "DSCasts.h"

#ifdef __WXMSW__
	long random() {
		return 0;
	}

	void srandom(unsigned int seed) {
	}
#endif

DynSoft::DSSession::DSSession(
	DSQuery *query,
	const bool cleanUp,
	const wxString &tableName
) : DSData(
	query,
	cleanUp
) {
	this->tableName = tableName;
}

DynSoft::DSClassInfo DynSoft::DSSession::GetClassInfo() {
	return DynSoft::DSClassInfo(
		DS_SESSION_NAME,
		DS_SESSION_VERSION,
		DS_SESSION_AUTHOR
	);
}

wxString DynSoft::DSSession::CreateNewSession(const wxString &information) {
	wxString sessionId;

	do {
		sessionId = CreateNewSessionId();
	} while(Exists(sessionId));

	query->Prepare(_N(
		"INSERT INTO \":table:\" ("
		"	id, "
		"	ip, "
		"	useragent, "
		"	countdown, "
		"	information"
		") VALUES ("
		"	:id:, "
		"	:ip:, "
		"	:useragent:, "
		"	NOW(), "
		"	:information:"
		")"
	));
	query->Bind(_N("table"), tableName, false);
	query->Bind(_N("id"), sessionId);
	query->Bind(_N("ip"), DSCgi::GetVariable(_N("REMOTE_ADDR")));
	query->Bind(_N("useragent"), DSCgi::GetVariable(_N("HTTP_USER_AGENT")));
	if(!information.IsEmpty())
		query->Bind(_N("information"), information);
	else
		query->Bind(_N("information"), _N("NULL"), false);
	query->Exec(false);

	DeleteOld();

	return sessionId;
}

bool DynSoft::DSSession::Exists(const wxString &sessionId) {
	query->Prepare(_N("SELECT COUNT(*) AS counter FROM \":table:\" WHERE id = :id:"));
	query->Bind(_N("table"), tableName, false);
	query->Bind(_N("id"), sessionId);
	query->Exec();
	query->NextRecord();
	bool exists = query->GetField(_N("counter")).ToInt() > 0;
	query->FreeResult();
	return exists;
}

bool DynSoft::DSSession::IsValid(
	const wxString &sessionId,
	const bool actualize
) {
	query->Prepare(_N(
		"SELECT "
		"	COUNT(id) AS counter "
		"FROM "
		"	\":table:\" "
		"WHERE "
		"	id = :id: "
		"AND "
		"	EXTRACT('epoch' FROM countdown) + 21600 >= EXTRACT('epoch' FROM NOW()) "
		"AND "
		"	useragent = :useragent: "
		"AND "
		"	ip = :ip:"
	));
	query->Bind(_N("table"), tableName, false);
	query->Bind(_N("id"), sessionId);
	query->Bind(_N("ip"), DSCgi::GetVariable(_N("REMOTE_ADDR")));
	query->Bind(_N("useragent"), DSCgi::GetVariable(_N("HTTP_USER_AGENT")));
	query->Exec();
	query->NextRecord();
	bool active = query->GetField(_N("counter")).ToInt() > 0;
	query->FreeResult();

	if(actualize) {
		query->Prepare(_N("UPDATE sessions SET countdown = NOW() WHERE id = :id:"));
		query->Bind(_N("id"), sessionId);
		query->Exec(false);
	}

	DeleteOld();

	return active;
}

wxString DynSoft::DSSession::CreateNewSessionId() {
	char buffer [11];
	srandom(time(NULL));
	srandom(random());
	snprintf(buffer, 11, "%ld%ld", random(), random());
	return DSCasts::ToString(buffer);
}

bool DynSoft::DSSession::DeleteOld() {
	query->Prepare(_N(
		"DELETE FROM "
		"	\":table:\" "
		"WHERE "
		"	EXTRACT('epoch' FROM countdown) + 21600 < EXTRACT('epoch' FROM NOW())"
	));
	query->Bind(_N("table"), tableName, false);
	return query->Exec(false);
}

bool DynSoft::DSSession::RemoveSession(const wxString &sessionId) {
	query->Prepare(_N("DELETE FROM \":table:\" WHERE id = :id:"));
	query->Bind(_N("table"), tableName, false);
	query->Bind(_N("id"), sessionId);
	return query->Exec(false);
}

wxString DynSoft::DSSession::GetInformation(const wxString &sessionId) {
	wxString information;
	
	query->Prepare(_N("SELECT information FROM \":table:\" WHERE id = :id:"));
	query->Bind(_N("table"), tableName, false);
	query->Bind(_N("id"), sessionId);
	query->Exec();
	if(query->NextRecord())
		information = query->GetField(_N("information")).ToString();
	query->FreeResult();

	return information;
}

bool DynSoft::DSSession::SetInformation(
	const wxString &sessionId,
	const wxString &information
) {
	query->Prepare(_N("UPDATE \":table:\" SET information = :information: WHERE id = :id:"));
	query->Bind(_N("table"), tableName, false);
	if(!information.IsEmpty())
		query->Bind(_N("information"), information);
	else
		query->Bind(_N("information"), _N("NULL"), false);
	query->Bind(_N("id"), sessionId);
	return query->Exec(false);
}

File: src/lib/php/DSSession.php
<?php

require_once "DSData.php";
require_once "DSClassInfo.php";

define("DS_SESSION_VERSION", "0.0.1");
define("DS_SESSION_NAME",    "DSSession");
define("DS_SESSION_AUTHOR",  "Thorsten Geppert (thorsten@ext2fs.de)");

class DSSession extends DSData {

	public function __construct(DSQuery $query) {
		parent::__construct($query);
	}

	public function GetClassInfo() {
		return new DSClassInfo(
			DS_SESSION_NAME,
			DS_SESSION_VERSION,
			DS_SESSION_AUTHOR
		);
	}

	public function CreateNewSession($information = "") {
		$sessionId = "";

		do {
			$sessionId = $this->CreateNewSessionId();
		} while($this->Exists($sessionId));

		$this->query->Prepare("
			INSERT INTO sessions (
				id, 
				ip, 
				useragent, 
				countdown, 
				information
			) VALUES (
				:id:, 
				:ip:, 
				:useragent:, 
				NOW(), 
				:information:
			)
		");
		$this->query->Bind("id", $sessionId);
		$this->query->Bind("ip", $_SERVER["REMOTE_ADDR"]);
		$this->query->Bind("useragent", $_SERVER["HTTP_USER_AGENT"]);
		$this->query->Bind("information", $information);
		$this->query->Exec(false);

		$this->DeleteOld();

		return $sessionId;
	}

	public function Exists($sessionId) {
		$this->query->Prepare("SELECT COUNT(*) AS counter FROM sessions WHERE id = :id:");
		$this->query->Bind("id", $sessionId);
		$this->query->Exec();
		$this->query->NextRecord();
		$exists = $this->query->GetField("counter") > 0;
		$this->query->FreeResult();
		return $exists;
	}

	public function IsValid($sessionId, $actualize = true) {
		$this->query->Prepare("
			SELECT 
				COUNT(id) AS counter 
			FROM 
				sessions 
			WHERE 
				id = :id: 
			AND 
				EXTRACT('epoch' FROM countdown) + 21600 >= EXTRACT('epoch' FROM NOW()) 
			AND 
				useragent = :useragent: 
			AND 
				ip = :ip:
		");
		$this->query->Bind("id", $sessionId);
		$this->query->Bind("ip", $_SERVER["REMOTE_ADDR"]);
		$this->query->Bind("useragent", $_SERVER["HTTP_USER_AGENT"]);
		$this->query->Exec();
		$this->query->NextRecord();
		$active = $this->query->GetField("counter") > 0;
		$this->query->FreeResult();

		if($actualize == true) {
			$this->query->Prepare("UPDATE sessions SET countdown = NOW() WHERE id = :id:");
			$this->query->Bind("id", $sessionId);
			$this->query->Exec(false);
		}

		$this->DeleteOld();

		return $active;
	}

	public function CreateNewSessionId() {
		return rand(11111111, 99999999);
	}

	protected function DeleteOld() {
		return $this->query->Query("
				DELETE FROM 
					sessions 
				WHERE 
					EXTRACT('epoch' FROM countdown) + 21600 < EXTRACT('epoch' FROM NOW())
			",
			false
		);
	}

	public function RemoveSession($sessionId) {
		$this->query->Prepare("DELETE FROM sessions WHERE id = :id:");
		$this->query->Bind("id", $sessionId);
		return $this->query->Exec(false);
	}

	public function GetInformation($sessionId) {
		$information = "";
		
		$this->query->Prepare("SELECT information FROM sessions WHERE id = :id:");
		$this->query->Bind("id", $sessionId);
		$this->query->Exec();
		if($this->query->NextRecord())
			$information = $this->query->GetField("information");
		$this->query->FreeResult();

		return $information;
	}

	public function SetInformation($sessionId, $information) {
		$this->query->Prepare("UPDATE sessions SET information = :information: WHERE id = :id:");
		$this->query->Bind("information", $information);
		$this->query->Bind("id", $sessionId, false);
		return $this->query->Exec(false);
	}

}

?>
